python/pyiceberg/transforms.py

```diff
@@ -96,6 +127,38 @@ def dedup_name(self) -> str:
     def __str__(self) -> str:
         return self.__root__
 
+    def __eq__(self, other: Any) -> bool:
+        if isinstance(other, Transform):
+            return self.__root__ == other.__root__
+        return False
+
+
+class UnboundTransform(Transform):
```

Comment:
Looking at the Java side a little more, I think there's an alternative to making this `UnboundTransform` change. The Java implementation really only needs the source type for the `apply` method. I think you could solve this problem by changing the `Transform` API. Instead of having `apply` on `Transform`, you could put `apply` on a `Transformer` class, then add `forType(t: IcebergType) -> Transformer` to `Transform`. That may be cleaner in the long term, although I think this works for now.

Reply:
If I understand correctly, this is what we already have for each of the classes of Transforms:
```python
def bucket(source_type: IcebergType, num_buckets: int) -> BaseBucketTransform:
    if type(source_type) in {IntegerType, LongType, DateType, TimeType, TimestampType, TimestamptzType}:
        return BucketNumberTransform(source_type, num_buckets)
    elif isinstance(source_type, DecimalType):
        return BucketDecimalTransform(source_type, num_buckets)
    elif isinstance(source_type, StringType):
        return BucketStringTransform(source_type, num_buckets)
    elif isinstance(source_type, BinaryType):
        return BucketBytesTransform(source_type, num_buckets)
    elif isinstance(source_type, FixedType):
        return BucketBytesTransform(source_type, num_buckets)
    elif isinstance(source_type, UUIDType):
        return BucketUUIDTransform(source_type, num_buckets)
    else:
        raise ValueError(f"Cannot bucket by type: {source_type}")
```
This is actually called on the `.bind()` method by the transform:
```python
bind_func = functools.partial(bucket, num_buckets=BaseBucketTransform.parse(v))
```
As soon as the source_type is known, we can invoke the method and get the actual implementation.

Reply:
Yes, this definitely works.

My point was just that we could structure transforms slightly differently, so that we don't need to bind in most situations. The only time we actually need to get the concrete source type for a transform is when we are calling `apply`. The other method signatures pass a type in. For example, `can_transform(IcebergType)` doesn't need to be specific to a `LongType` -- we can write a version for `BucketTransform` instead of for `BucketNumberTransform`.

In that case, we'd change the code so that instead of calling `apply` on `BucketNumberTransform`, we'd call `BucketTransform.bind` to get `Function[int, int]` that implements the hash and bucketing logic.

That refactor may end up being easier to work with in the end because there would be far fewer places where we'd need to bind the transform to a specific type.

Reply:
It may not be something that we want to do in this PR. We can follow up to change it, probably

Reply:
I have a working version, which make the code much simpler, which I like. I think it makes sense to do it in this PR, since it will remove 80% of the tests I already wrote :) I think the code is much nicer because we can omit the bind() stuff, including all the copying of the object.

Reply:
Sounds great!